### PR TITLE
payg: Implement an unlock dialog for PAY and integrate it with the lock screen

### PR DESCRIPTION
**Pull request by mariospr**
This will communicate with the com.endlessm.Payg1 D-Bus service and allow unlocking machines where the current PAYG has expired, as well as react to changes in the PAYG state of the machine.

https://phabricator.endlessm.com/T21608

### CODE REVIEW

File: js/misc/paygManager.js [hunk 1]
```diff
@@ -125,6 +125,14 @@ var PaygManager = new Lang.Class({
     get expiryTime() {
         return this._expiryTime;
     },
+
+    get isLocked() {
+        if (!this.enabled)
+            return false;
+
+        return this._expiryTime < (GLib.get_real_time() / 1e6)
```
Comment (pwithnall) [inline js/misc/paygManager.js]: I assume there isn’t a `GLib.USEC_PER_SEC` constant available?

Comment (pwithnall) [reply js/misc/paygManager.js]: Technically, this should be `<=`, not `<`.

File: js/misc/paygManager.js [hunk 2]
```diff
@@ -43,6 +43,22 @@ const EOS_PAYG_IFACE = '<node> \
 </interface> \
 </node>';
 
+var PaygErrorDomain = GLib.quark_from_string('payg-manager-error');
```
Comment (pwithnall) [inline js/misc/paygManager.js]: Nitpick: I think the variable name (`PaygErrorDomain`) should match the quark string `(`payg-manager-error`) to avoid any possible confusion in future. So this should have a quark string of `payg-error`, not `payg-manager-error`.

File: js/misc/paygManager.js [hunk 3]
```diff
@@ -110,8 +129,11 @@ var PaygManager = new Lang.Class({
         this.emit('code-expired');
     },
 
-    addCode: function(code) {
-        this._proxy.AddCodeRemote(code);
+    addCode: function(code, callback) {
+        this._proxy.AddCodeRemote(code, (result, proxyError) => {
+            if (callback)
+                callback(proxyToPaygError(proxyError));
```
Comment (pwithnall) [inline js/misc/paygManager.js]: `proxyToPaygError()` doesn’t seem to be defined anywhere.

File: po/POTFILES.in [hunk 4]
```diff
@@ -9,6 +9,7 @@ js/extensionPrefs/main.js
 js/gdm/authPrompt.js
 js/gdm/loginDialog.js
 js/gdm/util.js
+js/misc/paygManager.js
```
Comment (pwithnall) [inline po/POTFILES.in]: Why is this being added to `POTFILES.in` in this commit? The commit doesn’t add any translatable strings, as far as I can see.

File: js/ui/paygUnlockDialog.js [hunk 5]
```diff
@@ -0,0 +1,395 @@
+// -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Should this file have a copyright header?

File: js/ui/paygUnlockDialog.js [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+const PaygManager = imports.misc.paygManager;
+
+const Animation = imports.ui.animation;
+const Main = imports.ui.main;
+const Monitor = imports.ui.monitor;
+const ShellEntry = imports.ui.shellEntry;
+const Tweener = imports.ui.tweener;
+
+// The timeout before going back automatically to the lock screen (in seconds)
+const IDLE_TIMEOUT = 2 * 60;
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: smcv taught me a good approach for dealing with units: encode them in variable name, then you never have to look back somewhere else to remind yourself whether something is in seconds, microseconds or milliseconds. So this would be `const IDLE_TIMEOUT_SECS = …`.

File: js/ui/paygUnlockDialog.js [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+
+const CODE_ENTRY_REQUIRED_LENGTH = 8;
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: Similarly, it might be clearer if this was suffixed `_CHARS`.

File: js/ui/paygUnlockDialog.js [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+
+var SPINNER_ICON_SIZE = 16;
+var SPINNER_ANIMATION_DELAY = 1.0;
+var SPINNER_ANIMATION_TIME = 0.3;
+
+var MESSAGE_FADE_OUT_ANIMATION_TIME = 0.5;
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: Same for all the above.

File: js/ui/paygUnlockDialog.js [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+
+var UnlockStatus = {
+    NOT_VERIFYING: 0,
+    VERIFYING: 1,
+    FAILED: 2,
+    SUCCEEDED: 3
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: Add a trailing comma to prevent diff noise in future?

File: js/ui/paygUnlockDialog.js [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        if (!GLib.unichar_isprint(character))
+            return Clutter.EVENT_PROPAGATE;
+
+        // Don't allow inserting more digits than required.
+        if (this._code.length >= CODE_ENTRY_REQUIRED_LENGTH)
+            return Clutter.EVENT_STOP;
+
+        // Allow digits only
+        if (GLib.unichar_isdigit(character))
+            this.clutter_text.insert_unichar(character);
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: It seems a bit awkward that every `if` statement in this function causes an early `return` except this one. Refactor?

File: js/ui/paygUnlockDialog.js [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+        this.clutter_text.connect('text-changed', this._onTextChanged.bind(this));
+    },
+
+    _onCapturedEvent: function(textActor, event) {
+        if (event.type() != Clutter.EventType.KEY_PRESS)
+            return Clutter.EVENT_PROPAGATE;
+
+        let character = event.get_key_unicode();
+
+        // We only support printable characters.
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Question: What about hitting ‘enter’? Should that be handled explicitly here, or does Clutter handle that already in some other code path?

File: js/ui/paygUnlockDialog.js [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+        this._entry = null;
+        this._errorMessage = null;
+        this._cancelButton = null;
+        this._nextButton = null;
+        this._spinner = null;
+        this._cancelled = false;
+
+        this._verificationStatus = UnlockStatus.NOT_VERIFYING;
+
+        // Clear the clipboard to make sure nothing can be copied into the entry.
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Will this clear the clipboard in the user’s session? I’d be pretty annoyed if I had something on the clipboard, then the computer locked, and then I lost what I had.

File: js/ui/paygUnlockDialog.js [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+        promptBox.connect('key-press-event', (actor, event) => {
+            if (event.get_key_symbol() == Clutter.KEY_Escape)
+                this._onCancelled();
+
+            return Clutter.EVENT_PROPAGATE;
+        });
+        this.actor.add_child(promptBox);
+
+        let titleLabel = new St.Label({ style_class: 'unlock-dialog-payg-title',
+                                        text: _("Your subscription has expired!"),
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: Unless this was mandated by the designers, I’d drop the `!`. It seems a bit excited and unprofessional for something which is essentially about to ask the user for money.

File: js/ui/paygUnlockDialog.js [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+        let promptLabel = new St.Label({ style_class: 'unlock-dialog-payg-label',
+                                         text: _("Enter a new code to unlock the computer:"),
+                                         x_align: Clutter.ActorAlign.START });
+        promptBox.add_child(promptLabel);
+
+        this._entry = new PaygUnlockCodeEntry();
+        promptBox.add_child(this._entry);
+
+        this._errorMessage = new St.Label({ opacity: 0,
+                                       styleClass: 'unlock-dialog-payg-message' });
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: Indentation problem here, I think.

File: js/ui/paygUnlockDialog.js [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+                                             reactive: true,
+                                             can_focus: true,
+                                             label: _("Cancel"),
+                                             x_align: St.Align.START,
+                                             y_align: St.Align.END });
+        buttonsBox.add_child(this._cancelButton);
+
+        let buttonSpacer = new St.Widget({ layout_manager: new Clutter.BinLayout(),
+                                             x_expand: true,
+                                             x_align: Clutter.ActorAlign.END });
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: Indentation problem here, I think.

File: js/ui/paygUnlockDialog.js [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+
+    _updateSensitivity: function() {
+        this._updateNextButtonSensitivity();
+        this._entry.setEnabled(this._verificationStatus != UnlockStatus.VERIFYING);
+    },
+
+    _setErrorMessage: function(message) {
+        if (message) {
+            this._errorMessage.text = message;
+            this._errorMessage.opacity = 255;
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: I assume you’re hiding/showing `this._errorMessage` using opacity rather than `.show()`/`.hide()`, to avoid it causing relayouts?

File: js/ui/paygUnlockDialog.js [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+
+    _reset: function() {
+        this._stopSpinning();
+        this._entry.reset();
+        this._updateSensitivity();
+    },
+
+    _showError: function(error) {
+        if (error.matches(PaygManager.PaygErrorDomain, PaygManager.PaygError.DISABLED))
+            this._setErrorMessage(_("Pay-as-you-go is not enabled in this machine"));
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: We probably want to bypass the `PaygUnlockDialog` altogether if something returns the `DISABLED` error, rather than displaying an error message. The idea is that if this error is returned, the user probably isn’t even aware that PAYG is a concept.

File: js/ui/screenShield.js [hunk 18]
```diff
@@ -580,6 +580,47 @@ var ScreenShield = new Lang.Class({
...
 
         this._syncInhibitor();
+
+        Main.paygManager.connect('code-expired', () => { this.lock(true) });
+        Main.paygManager.connect('expiry-time-changed', () => {
+            let userManager = AccountsService.UserManager.get_default();
+            let user = userManager.get_user(GLib.get_user_name());
+
+            // A new valid code has been introduced but the machine is
+            // not unlocked yet -> Make sure we stay in the locked state.
```
Comment (pwithnall) [inline js/ui/screenShield.js]: This shouldn’t happen, I think (whenever a code is successfully verified, the `ExpiryTime` advertised by the daemon is set to now + the period of the code, so it should always be in the future). Is this branch just as a safety net?

File: js/misc/paygManager.js (hunk 1 above)
Comment (mariospr) [reply js/misc/paygManager.js]: > I assume there isn’t a GLib.USEC_PER_SEC constant available?

Actually there is, will change that.

> Technically, this should be <=, not <.

Correct

File: js/misc/paygManager.js (hunk 2 above)
Comment (mariospr) [reply js/misc/paygManager.js]: Ok

File: js/misc/paygManager.js (hunk 3 above)
Comment (mariospr) [reply js/misc/paygManager.js]: Oops! Somehow I messed up right before pushing my branch, since this belongs to a very old version that I no longer use. I guess I didn't notice because I was using a custom implementation of the manager for development, and the real thing (integration with the D-Bus daemon is not really tested). 

Will fix that, of course

File: po/POTFILES.in (hunk 4 above)
Comment (mariospr) [reply po/POTFILES.in]: Another leftover from a previous version where I did actually had translatable strings here, in the -now gone- `proxyToPaygError()` function.

File: js/ui/paygUnlockDialog.js (hunk 5 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Yes, it should. I slipped here due to starting off unlockDialog.js, which does not have one, but I will add it

File: js/ui/paygUnlockDialog.js (hunk 6 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Sounds good

File: js/ui/paygUnlockDialog.js (hunk 7 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Noted

File: js/ui/paygUnlockDialog.js (hunk 8 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Ok. Also, I'll remove the last one, which is unused (comes from unlockDialog.js)

File: js/ui/paygUnlockDialog.js (hunk 9 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Ok

File: js/ui/paygUnlockDialog.js (hunk 13 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: This was basically a placeholder, meant to be replaced with whatever we agreed upon, and the `!` was a reminder of that temporary nature. But I will remove it, in case it becomes the final thing.

File: js/ui/paygUnlockDialog.js (hunk 14 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Ah, yes

File: js/ui/paygUnlockDialog.js (hunk 15 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: WTF... ok

File: js/ui/paygUnlockDialog.js (hunk 16 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Actually, I'm not 100% sure, since I'm inheriting this "technique" from the unlockDialog.js module, although I thought exactly the same at the time.

File: js/ui/paygUnlockDialog.js (hunk 17 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Well, actually that is the case already (this dialog won't ever show up if PAYG is disabled) so this can't really ever happen. I had it mostly for testing purposes, but I'll remove it.

File: js/ui/screenShield.js (hunk 18 above)
Comment (mariospr) [reply js/ui/screenShield.js]: It is just that, yes

**Comment by mariospr** [comment]
> Review of “payg: Make the PaygManager a permanent component, not a loadable one”: +1
> 
> Although this reminds me that eos-paygd needs to be added a run-time dependency in the Debian > packaging at some point.

Sure, I'll add that on a separate PR for the debian-master branch

> Review of commit: “st-private: Add support for letter-spacing in ClutterText based StWidgets”: +1, although it would be better if you could trim the commit title.

I'll wait for the review upstream first if you don't mind. See https://gitlab.gnome.org/GNOME/gnome-shell/issues/154

> Review of commit “st-entry: Set text-related CSS properties on the internal ClutterText”: +1, though st and Clutter are definitely not my area of expertise.

Let's see what upstream has to say in https://gitlab.gnome.org/GNOME/gnome-shell/issues/154

> Review of commit “payg: Visual changes to the code entry widget to make it stand out”: +1 cocktail

Note that this is a proposal of mine, that needs buy-in from the design team (although I showed it to Robin and I think she was kind of ok with it)

> Review of commit “payg: Integrate the unlocking dialog for PAYG in the screen shield”: +1 with one question. I found it a bit hard to follow the code (a lot of state management which only makes sense when reading the wider context), so I think this one would benefit more from actual testing (which I know you’ve done) than code review.

Yes, it's hard. Also, it definitely needs more testing since I just spotted a few issues that I couldn't reproduce when using my mock implementation of PaygManager, so please don't merge anything just yet, even if you see fixup commits (which I'd rather squash myself before pushing).

File: js/ui/paygUnlockDialog.js (hunk 13 above)
Comment (pwithnall) [reply js/ui/paygUnlockDialog.js]: Ah, I see. I would have left a comment like `/* FIXME: Put final UI strings in place */` or summat.

File: js/ui/paygUnlockDialog.js (hunk 17 above)
Comment (pwithnall) [reply js/ui/paygUnlockDialog.js]: Leave an assertion in its place? Then it’s obvious why the code _doesn’t_ handle `PaygError.DISABLED`.

Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: I'm not sure if the equivalent to  `g_assert_not_reached()` is available to GJS, and in any case I feel like we should not crash in that case so I'm adding an `else` branch that would set the `Unknown error` message in the UI, along with a comment in the code to make it explicit that we don't consider any other error.

File: js/ui/paygUnlockDialog.js [hunk 19]
```diff
@@ -20,22 +38,22 @@ const Monitor = imports.ui.monitor;
 const ShellEntry = imports.ui.shellEntry;
 const Tweener = imports.ui.tweener;
 
-// The timeout before going back automatically to the lock screen (in seconds)
-const IDLE_TIMEOUT = 2 * 60;
+const ONE_SECOND = 1000
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: I’d call this something like `MSEC_PER_SEC`, partly to fit in with GLib’s `USEC_PER_SEC`, and partly to make it clear that this is in milliseconds.

File: js/ui/paygUnlockDialog.js [hunk 20]
```diff
@@ -1,4 +1,22 @@
 // -*- mode: js; js-indent-level: 4; indent-tabs-mode: nil -*-
+//
+// Copyright (C) 2018 Endless Mobile, Inc.
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: ©?

File: js/ui/paygUnlockDialog.js [hunk 21]
```diff
@@ -319,13 +319,19 @@ var PaygUnlockDialog = new Lang.Class({
...
-        else if (error.matches(PaygManager.PaygErrorDomain, PaygManager.PaygError.CODE_ALREADY_USED))
+        } else if (error.matches(PaygManager.PaygErrorDomain, PaygManager.PaygError.CODE_ALREADY_USED)) {
             this._setErrorMessage(_("Code already used"));
-        else if (error.matches(PaygManager.PaygErrorDomain, PaygManager.PaygError.TOO_MANY_ATTEMPTS))
+        } else if (error.matches(PaygManager.PaygErrorDomain, PaygManager.PaygError.TOO_MANY_ATTEMPTS)) {
             this._setErrorMessage(_("Too many attempts"));
+        } else {
+            // We don't consider any other error here (and we don't consider DISABLED explicitly,
+            // since that should not happen), but still we need to show something to the user.
+            this._setErrorMessage(_("Unknown Error"));
```
Comment (pwithnall) [inline js/ui/paygUnlockDialog.js]: Nitpick: The Capitalisation Of The Error Message Is Inconsistent With The Error Messages Above.

**Comment by mariospr** [comment]
> Review of commit “payg: Make the initialization of the PaygManager synchronous”:
> 
> Typo ‘efore’ in the commit message

Ok

> The code looks OK, but I’m wondering about blocking. In C, I’d make PaygManager itself asynchronous (GAsyncInitable) and yield construction of the other objects (like the screenshield) on asynchronous initialisation of the PaygManager. That means we don’t block the main loop while waiting for it to be initialised. However, I don’t know if that’s the right paradigm for JS (and even if it is, it’s a fair amount of refactoring).

I'm not very comfortable with blocking here either, but at the same time I wanted to be the least intrusive possible with the code in `main.js`, which is why I tried this approach. That said, I think it's worth looking for an alternative solution, so I'll spend a bit more time looking into that.

**Comment by mariospr** [comment]
> Review of commit “fixup! payg: Add an isLocked attribute to the PaygManager”: +1, but it’s going to have to be reworked anyway since eos-paygd now operates in seconds (sorry).

Yes

File: js/ui/paygUnlockDialog.js (hunk 19 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: Sounds good. I used ONE_SECOND for consistency with https://github.com/endlessm/gnome-shell/blob/519a389dae2f1b8c7fc28b944b4f6966f8dcb359/js/ui/windowManager.js#L46, but I like more your proposed alternative.

File: js/ui/paygUnlockDialog.js (hunk 21 above)
Comment (mariospr) [reply js/ui/paygUnlockDialog.js]: You. Are. right (hah!)

**Comment by mariospr** [comment]
> I'm not very comfortable with blocking here either, but at the same time I wanted to be the least intrusive possible with the code in main.js, which is why I tried this approach. That said, I think it's worth looking for an alternative solution, so I'll spend a bit more time looking into that.

I reverted that patch on top of my branch and then pushed 2 more commits that I believe (I tested it) achieve the same result without blocking.

File: js/misc/paygManager.js [hunk 22]
```diff
@@ -86,18 +88,23 @@ var PaygManager = new Lang.Class({
...
         } catch (e) {
             logError(e, "Error while constructing D-Bus proxy for " + EOS_PAYG_NAME);
-            return;
         }
 
-        this._setEnabled(this._proxy.Enabled);
-        this._setExpiryTime(this._proxy.ExpiryTime);
+        if (success) {
+            this._enabled = this._proxy.Enabled;
+            this._expiryTime = this._proxy.ExpiryTime;
```
Comment (pwithnall) [inline js/misc/paygManager.js]: Why not call `this._setEnabled()` and `this._setExpiryTime()`? Are you deliberately wanting to avoid emitting their signals before `initialized` is emitted?

File: js/ui/main.js [hunk 23]
```diff
@@ -241,7 +241,17 @@ function _initializeUI() {
...
-            screenShield.showDialog();
+            // We can't show the login dialog (which is managed by the
+            // screenshield) until the PaygManager is initializd, since
+            // we need to check whether the machine is PAYG-locked first.
+            if (paygManager.initialized) {
+                screenShield.showDialog();
+            } else {
+                let paygManagerId = paygManager.connect('initialized', () => {
+                    screenShield.showDialog();
+                    paygManager.disconnect(paygManagerId);
```
Comment (pwithnall) [inline js/ui/main.js]: What’s going to happen in the interim, while waiting for this async operation to finish?

Comment (mariospr) [reply js/ui/main.js]: Nothing. The screenshield is created already way before this async operation is finished, patiently waiting for that to happen so that `showDialog()` can be called.

Comment (mariospr) [reply js/ui/main.js]: If you mean "visually", then the answer is again "nothing": the login screen is actually constructed and shown from showDialog(), so nothing will be shown until its called, which is equivalent to the login screen (or the unlock dialog) to maybe taking a few extra miliseconds to show up.

File: js/misc/paygManager.js (hunk 22 above)
Comment (mariospr) [reply js/misc/paygManager.js]: YES, otherwise that confuses the shell on the way the screenshield handles the initial state, so I'd rather avoid that emission if you don't mind.

Comment (pwithnall) [reply js/misc/paygManager.js]: Add a comment in the code about that, and then I think we’re good to merge.
